pkgs/top-level/all-packages.nix

```diff
@@ -25753,9 +25753,9 @@ with pkgs;
   ### DEVELOPMENT / GO
 
   # the unversioned attributes should always point to the same go version
-  go = go_1_21;
-  buildGoModule = buildGo121Module;
-  buildGoPackage = buildGo121Package;
+  go = go_1_22;
```

Comment:
Why? Right now you cannot use the 1.22 toolchain for development, as tools like gopls, gotools and golangci-lint won't work if compiled with an older major release.

Reply:
I once got a review ([link](https://github.com/NixOS/nixpkgs/pull/250345#discussion_r1299381294)), that we wait for the first point release instead of a big major bump.
e.g if we are moving from go1.21 -> go1.22; we should prefer go1.22.1 instead of go1.22.0

Reply:
I think since its for unstable, I think not waiting for 1.22.1 is ok ( my opinion )

sorry to switch threads, my understanding is from 1.21 go version listed in go.mod is treated as requirement. How does this work ? did nixpkgs somehow changed that behavior. Shouldn't all packages listing go 1.21 in go.mod stop building ?

Reply:
> I once got a review ([link](https://github.com/NixOS/nixpkgs/pull/250345#discussion_r1299381294)), that we wait for the first point release instead of a big major bump. e.g if we are moving from go1.21 -> go1.22; we should prefer go1.22.1 instead of go1.22.0
> 
> /cc @zowoq

Since this wait is causing breakage in the toolchain, I strongly urge that unless we have a really good reason for waiting, not to wait.
@mfrw

Reply:
> my understanding is from 1.21 go version listed in go.mod is treated as requirement. How does this work ? did nixpkgs somehow changed that behavior. Shouldn't all packages listing go 1.21 in go.mod stop building ?

@kirillrdy To my understanding, the go version in go.mod is not an exact requirement and never was. It describes the _feature set_ needed to build this Go module. Starting from Go 1.21, this is strictly enforced for the module and its transitive dependencies. From the [release notes](https://tip.golang.org/doc/go1.21#tools):

> Go 1.21 now reads the go line in a go.work or go.mod file as a strict minimum requirement

Reply:
`Since this wait is causing breakage in the toolchain`

@prmadev can you clarify what breakage in toolchain is this causing ? any package if required can use 1.22. this PR is about making 1.22 default

Reply:
I agree; that we should not wait for go1.22.1
Although, please do not treat this as an authoritative comment, I am not an expert, just trying to learn.

Reply:
so nixpkgs-review got stuck, which probably explains why it's safer to wait for 1.22.1 release, otherwise we don't know what breaks until it lands in staging-next

> Since this wait is causing breakage in the toolchain

can we open an issue for this ? I've switched my work project to `go_1_22` ,and `gopls` and `golangci-lint` don't have any issues

Reply:
I install go_1_22 and gopls. In a repository with `go 1.22` in go.mod, gopls is not work. Please, upgrade go to 1.22.

Reply:
I've changed go.mod to 1.22 and still working.

@nanozuki  can we open separate issue for this ?

Reply:
Opening another issue is OK. But I don't think it's acceptable for a language like go, which emphasizes compatibility, to skip a major release and not upgrade it.

Reply:
nixpkgs is not going to skip a major release, its just for nixpkgs is a huge project and a lot of things need to be checked/fixed before changing default version, as for users/consumers of go its easy to get version of tools eg gopls to use different version of go

Reply:
This repository: https://github.com/sumnerevans/nixpkgs-290212 is proof that this affects gopls. The repository has a flake and a very basic Go project with an error that should be detected by gopls. If you clone the repository, and then run the following (paths redacted):

```
$ nix develop

[...]$ which gopls
/nix/store/sa9fczlr6rcip8inlb9z6l3hd2hy9gs2-gopls-0.15.1/bin/gopls

[...]$ cat go.mod
module 290212

go 1.22

[...]$ gopls check main.go
Log: Loading packages...
Info: Finished loading packages.
Log: Running...
Info: completed

[...]$ vim go.mod

[...]$ cat go.mod
module 290212

go 1.21

[...]$ gopls check main.go
Log: Loading packages...
Info: Finished loading packages.
Log: Running...
Info: completed
.../main.go:4:6-7: x declared and not used

[...]$
```

You can see that in the first case (when the `go.mod` specifies 1.22) no errors are shown, but when I changed it to 1.21 in the `go.mod`, it returns an error.

However, if I run gopls installed via `go install golang.org/x/tools/gopls@latest`, then I get the the `x declared and not used` error when 1.22 is specified in `go.mod`.

Reply:
I can reproduce the issue, thanks for that @sumnerevans 
lets move discussion to https://github.com/NixOS/nixpkgs/issues/293180

Reply:
> I highly doubt that this is related to the update. There are some really big builds affected by updating Go

those two sentences contradict each other

Reply:
s/affected/triggered to rebuild/

Reply:
even with substituted words

> I highly doubt that this is related to the update.

"update" here, i assume means changing go = go_1_21 to go_1_22.

> There are some really big builds triggered to rebuild by updating Go

rebuilds are caused by this update, so I am not sure what is the connection between two sentences

let's link back to original comment `so nixpkgs-review got stuck` which is what makes harder to estimate fallout from bumps likes this.

hopefully this clears things up, if not I am happy to clarify further

Reply:
Go 1.22.1 is in PR:
- https://github.com/NixOS/nixpkgs/pull/293580